identity/src/main/java/com/android/identity/internal/Util.kt

```diff
@@ -0,0 +1,2209 @@
+        val protectedHeadersMap: MapBuilder<CborBuilder> = protectedHeaders.addMap()
+        protectedHeadersMap.put(COSE_LABEL_ALG, signatureAlgorithm.coseAlgorithmIdentifier.toLong())
+        val protectedHeadersBytes = cborEncode(protectedHeaders.build().get(0))
+        val toBeSigned = coseBuildToBeSigned(protectedHeadersBytes, data, detachedContent)
+        val derSignature: ByteArray =
+            secureArea.sign(alias, signatureAlgorithm, toBeSigned, keyUnlockData)
+        val coseSignature = signatureDerToCose(derSignature, 32) // TODO: infer from alias
+        val builder = CborBuilder()
+        val array: ArrayBuilder<CborBuilder> = builder.addArray()
+        array.add(protectedHeadersBytes)
+        val unprotectedHeaders: MapBuilder<ArrayBuilder<CborBuilder>> = array.addMap()
+        try {
+            if (certificateChain != null && certificateChain.size > 0) {
+                if (certificateChain.size == 1) {
+                    val cert = certificateChain.iterator().next()
+                    unprotectedHeaders.put(COSE_LABEL_X5CHAIN, cert.encoded)
+                } else {
+                    val x5chainsArray: ArrayBuilder<MapBuilder<ArrayBuilder<CborBuilder>>> =
+                        unprotectedHeaders.putArray(
+                            COSE_LABEL_X5CHAIN
+                        )
+                    for (cert in certificateChain) {
+                        x5chainsArray.add(cert.encoded)
+                    }
+                }
+            }
+        } catch (e: CertificateEncodingException) {
+            throw IllegalStateException("Error encoding certificate", e)
+        }
+        if (data == null || data.size == 0) {
+            array.add(SimpleValue(SimpleValueType.NULL))
+        } else {
+            array.add(data)
+        }
+        array.add(coseSignature)
+        return builder.build().get(0)
+    }
+
+    /**
+     * Note: this uses the default JCA provider which may not support a lot of curves, for
+     * example it doesn't support Brainpool curves. If you need to use such curves, use
+     * [.coseSign1Sign] instead with a
+     * Signature created using a provider that does have support.
+     *
+     * Currently only ECDSA signatures are supported.
+     *
+     * TODO: add support and tests for Ed25519 and Ed448.
+     */
+    @JvmStatic
+    fun coseSign1Sign(
+        key: PrivateKey,
+        algorithm: String, data: ByteArray?,
+        additionalData: ByteArray?,
+        certificateChain: Collection<X509Certificate>?
+    ): DataItem {
+        return try {
+            val s = Signature.getInstance(algorithm)
+            s.initSign(key)
+            coseSign1Sign(s, data, additionalData, certificateChain)
+        } catch (e: NoSuchAlgorithmException) {
+            throw IllegalStateException("Caught exception", e)
+        } catch (e: InvalidKeyException) {
+            throw IllegalStateException("Caught exception", e)
+        }
```

Comment:
I think the number of lines it takes to rewrite this in kotlin dsl is larger than having 2 catches thrown, this is a meritable tradeoff, either way works